rosette/rbl/rosette/primitives.rbl

```diff
@@ -0,0 +1,91 @@
+;;; TODO: Implement match? for other data structures
+(defOprn match?)
+(defPure Tuple (match? other) (= (self) other))
+
+(defOprn get)
+(defPure RblTable (get key)
+    (tbl-get (self) (->symbol key))
+)
+
+(defOprn getOrElse)
+(defPure RblTable (getOrElse & keys/elseValue)
+    (let*
+        [
+            ;;; TODO: Find way to deconstruct without reversing
+            [[elseValue & reverse_keys] (reverse keys/elseValue)]
+            [unreversed_keys (reverse reverse_keys)]
+            [final_value
+                (fold unreversed_keys
+                    (proc [key acc k]
+                        ;;; TODO: (null? key) is a proxy for checking if the unreversed_keys have exhausted.
+                        ;;; We need to handle when key is null.
+                        (if (null? key)
+                            acc
+                            (if (type? acc RblTable)
+                                (let [[value (tbl-get acc (->symbol key))]]
+                                    (if (= #absent value) elseValue (k value))
+                                )
+                                elseValue
+                            )
+                        )
+                    )
+                    (self))]
+        ]
+        final_value
+    )
+)
+
+(defOprn insert)
+(defPure RblTable (insert & keys/value)
+    ;;; Inserts (new RblTable) if a lookup is #absent when looping through keys
+    (let*
+        [
+            ;;; TODO: Find way to deconstruct without reversing
```

Comment:
I don't know enough of the syntax to really parse what's going on here.  What does reversing a table even do?

Reply:
Ping

Reply:
If you are addressing `[[elseValue & reverse_keys] (reverse keys/elseValue)]`, it isn't reversing a table. What is passed in as an argument is keys followed by an elseValue. For example the tuple

2, 3, 4, 5

for keys/elseValue means in Java (approx)

```
Table t1 = table.get(2)
Table t2 = t1.get(3)
Value v = t2.getOrElse(4, 5)
return v
```

What I would really love to do is

[[keys & elseValue] keys/elseValue]

but that would bind the first key to keys and then the tail of the the list to elseValue. Handling this on the compiler side so that passing in [keys, elseValue] as a tuple felt even less elegant.

Reply:
https://rchain.atlassian.net/browse/ROS-272